server/src/main/java/org/elasticsearch/gateway/ReplicaShardAllocator.java

```diff
@@ -311,36 +301,34 @@ public static Tuple<Decision, Map<String, NodeAllocationResult>> canBeAllocatedT
                 if (explain) {
                     madeDecision = decision;
                 } else {
-                    return Tuple.tuple(decision, null);
+                    return new PerNodeAllocationResult(decision, null);
                 }
             } else if (madeDecision.type() == Decision.Type.NO && decision.type() == Decision.Type.THROTTLE) {
                 madeDecision = decision;
             }
             if (explain) {
-                nodeDecisions.put(node.nodeId(), new NodeAllocationResult(node.node(), null, decision));
+                nodeDecisions.add(new NodeAllocationResult(node.node(), null, decision));
```

Comment:
No need to collect to to map as this is only used to iterate over all values later.